domain/network/state/subnet.go

```diff
@@ -55,149 +55,173 @@ func (st *State) UpsertSubnets(ctx context.Context, subnets []network.SubnetInfo
+	}
+	// Add the provider network id and its uuid to the
+	// provider_network table.
+	if err := tx.Query(
 		ctx,
 		insertSubnetProviderNetworkIDStmt,
-		pnUUID.String(),
-		subnet.ProviderNetworkId,
-	); err != nil {
-		return errors.Annotatef(err, "inserting provider network id %q for subnet %q", subnet.ProviderNetworkId, uuid)
+		ProviderNetwork{ProviderNetworkUUID: pnUUID.String(), ProviderNetworkId: subnetInfo.ProviderNetworkId},
+	).Run(); err != nil {
+		return errors.Annotatef(err, "inserting provider network id %q for subnet %q", subnetInfo.ProviderNetworkId, uuid)
 	}
 	// Insert the providerNetworkUUID into provider network to
 	// subnets mapping table.
-	if _, err := tx.ExecContext(
+	if err := tx.Query(
 		ctx,
 		insertSubnetProviderNetworkSubnetStmt,
-		pnUUID.String(),
-		uuid,
-	); err != nil {
-		return errors.Annotatef(err, "inserting association between provider network id %q and subnet %q", subnet.ProviderNetworkId, uuid)
+		ProviderNetworkSubnet{SubnetUUID: uuid, ProviderNetworkUUID: pnUUID.String()},
+	).Run(); err != nil {
+		return errors.Annotatef(err, "inserting association between provider network id %q and subnet %q", subnetInfo.ProviderNetworkId, uuid)
 	}
 
-	return addAvailabilityZones(ctx, tx, uuid, subnet)
+	return addAvailabilityZones(ctx, tx, uuid, subnetInfo)
 }
 
 // addAvailabilityZones adds the availability zones of a subnet if they don't exist, and
 // update the availability_zone_subnet table with the subnet's id.
-func addAvailabilityZones(ctx context.Context, tx *sql.Tx, subnetUUID string, subnet network.SubnetInfo) error {
-	retrieveAvailabilityZoneStmt := `
-SELECT uuid
+func addAvailabilityZones(ctx context.Context, tx *sqlair.TX, subnetUUID string, subnet network.SubnetInfo) error {
+	retrieveAvailabilityZoneStmt, err := sqlair.Prepare(`
+SELECT &M.uuid
 FROM   availability_zone
-WHERE  name = ?`
-	insertAvailabilityZoneStmt := `
+WHERE  name = $M.name`, sqlair.M{})
+	if err != nil {
+		return errors.Trace(err)
+	}
+	insertAvailabilityZoneStmt, err := sqlair.Prepare(`
 INSERT INTO availability_zone (uuid, name)
-VALUES (?, ?)`
-	insertAvailabilityZoneSubnetStmt := `
+VALUES ($M.uuid, $M.name)`, sqlair.M{})
+	if err != nil {
+		return errors.Trace(err)
+	}
+	insertAvailabilityZoneSubnetStmt, err := sqlair.Prepare(`
 INSERT INTO availability_zone_subnet (availability_zone_uuid, subnet_uuid)
-VALUES (?, ?)`
+VALUES ($M.availability_zone_uuid, $M.subnet_uuid)`, sqlair.M{})
+	if err != nil {
+		return errors.Trace(err)
+	}
 
 	for _, az := range subnet.AvailabilityZones {
 		// Retrieve the availability zone.
-		var azUUIDStr string
-		row := tx.QueryRowContext(ctx, retrieveAvailabilityZoneStmt, az)
-		err := row.Scan(&azUUIDStr)
+		m := sqlair.M{}
+		err := tx.Query(ctx, retrieveAvailabilityZoneStmt, sqlair.M{"name": az}).Get(m)
 		if err != nil && err != sql.ErrNoRows {
 			return errors.Annotatef(err, "retrieving availability zone %q for subnet %q", az, subnetUUID)
 		}
```

Comment:
```suggestion
		if err := tx.Query(ctx, retrieveAvailabilityZoneStmt, sqlair.M{"name": az}).Get(m); err != nil && err != sql.ErrNoRows {
			return errors.Annotatef(err, "retrieving availability zone %q for subnet %q", az, subnetUUID)
		}
```

Reply:
This one can't be in-lined because the err is checked for `sqlair.ErrNoRows` further down using the `err` var

Reply:
oh yes! my bad, sorry for that